[Batch table] `getDataForId` and `3DTILES_batch_table_hierarchy` support

mgermerie opened the pull request:
This PR adds : 

- A `BatchTable.getDataForId` method which returns all data in a batch table for a given batch `ID`.
- The support for the [`3DTILES_batch_table_hierarchy`](https://github.com/CesiumGS/3d-tiles/blob/main/extensions/3DTILES_batch_table_hierarchy/README.md) extension to enrich `BatchTable.getDataForId` with data eventually stored using this extension.

This PR should close #595.

**Here is an explanation on a few things I did :**

I added a [`BatchTableHierarchyExtension`](https://github.com/mgermerie/3DTilesRendererJS/blob/5660f778cafe7cf2fb14373f0371698f2c0ed960/src/utilities/BatchTableHierarchyExtension.js) class that allows parsing batch table data formatted using [`3DTILES_batch_table_hierarchy`](https://github.com/CesiumGS/3d-tiles/blob/main/extensions/3DTILES_batch_table_hierarchy/README.md) extension.

I defined a `parseBinArray` function in [`FeatureTable.js`](https://github.com/mgermerie/3DTilesRendererJS/blob/5660f778cafe7cf2fb14373f0371698f2c0ed960/src/utilities/FeatureTable.js#L3) file. It is simply a factorization of already existing code that allows parsing data from a binary batch table in [`BatchTableHierarchyExtension`](https://github.com/mgermerie/3DTilesRendererJS/blob/5660f778cafe7cf2fb14373f0371698f2c0ed960/src/utilities/BatchTableHierarchyExtension.js) without needing to rewrite this function.

I defined `BatchTable` in its own [`BatchTable.js`](https://github.com/mgermerie/3DTilesRendererJS/blob/5660f778cafe7cf2fb14373f0371698f2c0ed960/src/utilities/BatchTable.js) file, outside of [`FeatureTable.js`](https://github.com/mgermerie/3DTilesRendererJS/blob/5660f778cafe7cf2fb14373f0371698f2c0ed960/src/utilities/FeatureTable.js) file. This is just to prevent circular dependency. 
Otherwise, we would have the following :
```js
// In BatchTableHierarchyExtension.js file :
import { parseBinArray } from './FeatureTable.js';
```
```js
// In FeatureTable.js file :
import { BatchTableHierarchyExtension } from './BatchTableHierarchyExtension.js';
```

**Notes and questions :**

- I never wrote type definition files before so don't hesitate to mention if I'm missing some points in those I edited.
- If I understood well, everything that is defined within a type definition file describes what is in the public API (plus the exports in `index.js`). Should I add a `BatchTableHierarchyExtension.d.ts` file or would you rather have it not exposed ?
- I chose to register the `BatchTableHierarchyExtension` within `BatchTable` constructor, but wonder of different ways to do it : would it be better to register it within [`B3DMLoaderBase`](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/ac83abe6f82974ae2f2d4d146a99f33b406bd414/src/base/loaders/B3DMLoaderBase.js#L70), [`I3DMLoaderBase`](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/ac83abe6f82974ae2f2d4d146a99f33b406bd414/src/base/loaders/I3DMLoaderBase.js#L73) and [`PNTSLoaderBase`](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/ac83abe6f82974ae2f2d4d146a99f33b406bd414/src/base/loaders/PNTSLoaderBase.js#L69) (a bit like `3DTILES_draco_point_compression` extension is registered in [`PNTSLoader`](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/ac83abe6f82974ae2f2d4d146a99f33b406bd414/src/three/loaders/PNTSLoader.js#L39)) ?


Don't hesitate if you need more info on the PR !

gkjohnson commented on src/utilities/BatchTable.js:
```diff
@@ -0,0 +1,60 @@
+
+		}
+
+		for ( const extension in this.extensions ) {
+
+			target[ extension ] = this.extensions[ extension ].getDataForId( id );
+
+		}
```
What's the rationale here? Is it the case that _every_ extension will have `getDataForId`?

mgermerie replied:
Whoops! Indeed adding a check for `getDataForId` existence is a bug prevention I completely missed. I'll add that.

I know of a research lab that uses custom made 3D tiles extensions, but I don't know if they use some on batch tables (perhaps @jailln knows more of it). Anyway, `BatchTable.extensions` would need to be public, with an extension registration tool for `BatchTable` to allow adding custom extensions. So I agree that we will probably only deal with batch table hierarchy extension.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,199 @@
+
+function parseProperty( property, buffer, binOffset, length ) {
+
+	if ( ! ( property instanceof Object ) ) {
+
+		throw new Error( 'BatchTableHierarchyExtension: unexpected value found in extension properties.' );
```
Lets remove this - I think we have to rely on the input data being well-formed otherwise we will have these kinds of checks everywhere.

mgermerie replied:
Agreed, I'll remove that.

gkjohnson commented:
> everything that is defined within a type definition file describes what is in the public API

Right - I've just been using the d.ts files to define the "public" API. So no need to add the hierarchical BatchedTable there.

> I chose to register the BatchTableHierarchyExtension within BatchTable constructor, but wonder of different ways to do it 

I guess the thing to consider here is whether these classes should _always_ be included in a build causing larger file sizes. Ideally this kind of thing would be a user-registerable plugin but I think this is okay for now. It might be worth just adding a "TODO" by the extension initialization noting that it's worth considering a plugin in the future.

But realistically the B3DM, I3DM, BatchTable, etc classes from 3D Tiles 1.0 are all deprecated so it may make sense to move all of that functionality to a 1.0-backwards compatible plugin later, anyway, to make the core module smaller.

mgermerie commented on src/utilities/BatchTable.d.ts:
```diff
@@ -0,0 +1,24 @@
+		type : String | null
+	) : Number | String | ArrayBufferView;
+
+	getDataForId(
+		id: Number,
+		target?: Object,
+	) : Object;
```
No problem :) I'll rename it.

mgermerie commented on src/utilities/BatchTable.d.ts:
```diff
@@ -0,0 +1,24 @@
+
+	getKeys() : Array< String >;
+
+	getData(
+		key : String,
+		componentType : String | null,
+		type : String | null
+	) : Number | String | ArrayBufferView;
```
Agreed, batch tables work more on an ID basis (according to my understanding of specifications).

gkjohnson replied:
Great - do you mind adding a log warning the user that the function is deprecated if they use it, then? Like so:

```js
getData( ...args ) {

  console.warn( ... );
  return this._getData( ...args );

}
```

mgermerie replied:
Yep I'll add that.

mgermerie replied:
`BatchTable.getData` method being used only once and its code being a one liner, I did not define a `BatchTable._getData` method. Let me know if that suits you.

mgermerie commented on src/utilities/FeatureTable.js:
```diff
@@ -1,5 +1,77 @@
+			break;
+
+		default:
+			return 'typeError';
```
Agree with you, though we don't have access to the `key` the `parseBinArray` method was called to parse data from. This changes the message that was sent to users before this PR. Is it ok for you to throw an error with a less specific message (without specifying the `key`)?

gkjohnson replied:
I'm inclined to say we should just add a parameter to `parseBinArray` that takes the name of property so we can throw a nice human-understandable error. Anything else that comes to mind feels unnecessarily messy and redundant (validating types beforehand, catching and rethrowing an error).

mgermerie replied:
Okay, I'll add the parameter.

mgermerie commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,199 @@
+
+	get instancesIds() {
+
+		if ( ! this._cache.instancesIds ) {
+
+			this._cache.instancesIds = [];
+
+			const classCounter = {};
+			for ( const classId of this.classIds ) {
+
+				classCounter[ classId ] = classCounter[ classId ] ?? 0;
+				this._cache.instancesIds.push( classCounter[ classId ] );
+				classCounter[ classId ] ++;
+
+			}
+
+		}
```
I opted for the lazy generation so that instancing a `BatchTableHierarchyExtension` does almost nothing but storing the raw data. It saves user who does not care for the batch table in his/her data the cost of parsing batch table hierarchy data, however small that cost. This is inspired by what's done in `FeatureTable` and `BatchTable`, where data is only parsed when `getData` method is called.

I admit it complicates `BatchTableHierarchyExtension` code, with the cache logic. If you would rather have data parsed in constructor, I can change it accordingly.

gkjohnson replied:
I think we should opt for code simplicity to start and optimize later when needed. The code in `parseBinArray` is very lightweight and I don't think it should cause any issues.

mgermerie replied:
Understood, I'll parse data in constructor.

mgermerie commented:
I pushed specific commits for each modification you suggested to ease the review. I'll squash them before merging.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,123 @@
+
+		for ( const key in instances ) {
+
+			if ( Object.prototype.hasOwnProperty.call( instances, key ) ) {
```
From the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) - we should be using `Object.hasOwn`, instead.

gkjohnson replied:
Also is there a reason `instances` _wouldn't_ have "key" at this point? Is this check needed?

mgermerie replied:
Indeed, it's something I added without reflection in the `for ... in` loop to prevent iterating through prototype and inherited properties but that is not needed here since the object is just a parsed JSON object. I'll remove it.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,123 @@
+
+			}
+
+			const parentIds = this.parentIds.slice(
+				parentIdsOffset,
+				parentIdsOffset + parentCount,
+			);
+
+			for ( const parentId of parentIds ) {
```
Can we just use a normal for loop here instead of making a new array?

```js
for ( let i = 0; i < parentCount; i ++ ) {

  const parentId = parentIds[ parentIdsOffset + i ];
```

mgermerie replied:
Indeed, it is simpler. I changed it.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,123 @@
+
+	}
+
+	getDataFromId( id ) {
```
For consistency this function should take an optional `target = {}`, as well.

gkjohnson commented on src/utilities/BatchTable.js:
```diff
@@ -0,0 +1,67 @@
+
+			if ( extension.getDataFromId instanceof Function ) {
+
+				target[ extensionName ] = extension.getDataFromId( id );
```
Is there a reason we can't just set the hierarch fields directly onto the target object? Is there risk of name overlap between the extension and the original BatchTable values?

mgermerie replied:
You named the reason : nothing in the extension specifications prevents having classes that share the same name with any property outside the extension data.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,123 @@
+		const className = this.classes[ classId ].name;
+		const instanceId = this.instancesIds[ id ];
+
+		for ( const key in instances ) {
```
From the `_parseProperty` function above - it looks like `instances` can be a value like a string or number (see the `! ( property instanceof Object )`) and therefore couldn't be iterated over - what happens here in that case?

mgermerie replied:
It's an inconsistency in `_parseProperty` function. Actually, according to the spec, properties parsed using `_parseProperty` are `Array`, `Object` defining a location in binary part, or `undefined` but only for `parentCounts` and `parentIds` properties.
I changed `_parseProperty` and its usage a bit [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/pull/627/commits/e49fc92aa5b28451e4bd22b7ece458755ce02c1a) to make it more relevant to the spec.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,199 @@
+
+}
+
+export class BatchTableHierarchyExtension {
```
Thanks for the detailed explanation! I've left a few more comments now that I understand a bit more.

mgermerie commented on src/utilities/FeatureTable.js:
```diff
@@ -61,74 +133,10 @@ export class FeatureTable {
-				case 'DOUBLE':
-					data = new Float64Array( buffer, arrayStart, arrayLength );
-					break;
+			const data = parseBinArray( buffer, arrayStart, count, featureType, featureComponentType );
```
Whoops, added it.

mgermerie commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,123 @@
+
+			const arrayStart = binOffset + byteOffset;
+
+			return parseBinArray( buffer, arrayStart, propertyLength, 'SCALAR', componentType );
```
The only way an error that uses the property name can be thrown when `parseBinArray` is called from `BatchTableHierarchyExtension` is having a `componentType` in data that is neither `undefined` nor one of the types described [here](https://github.com/CesiumGS/3d-tiles/blob/87218a7c2030890b5200cf16cb466dbbd1480a20/extensions/3DTILES_batch_table_hierarchy/schema/3DTILES_batch_table_hierarchy.schema.json#L56-L86).

Currently in `BatchTableHierarchyExtension`, an `undefined` `componentType` defaults to `'UNSIGNED_SHORT'` and `type` is hard coded as `'SCALAR'` (following the specification). Therefore, I think it is quite unlikely that the error messages thrown in `parseBinArray` is ever thrown, except for some weirdly ill-formatted data.

Moreover `BatchTableHierarchyExtension._parseProperty` is parametrized with property values instead of name (which is not the case in `FeatureTable.getData`, the other place `parseBinArray` is called from). I made it that way so that any property can be parsed using `_parseProperty`, and not only properties that share the same parent object. We can thus parse `extensionHeader.classIds` with the same method we use to parse `extensionHeader.classes[ x ].instances[ someProperty ]`). We would therefore need to add a parameter to `BatchTableHierarchyExtension._parseProperty` just to pass the property name (which would be hard coded each time).

Therefore, I would suggest not to add the property name in this situation, though I'll add it if you still want it added :smile:

gkjohnson replied:
Got it - we should add something for an argument here or set the `propertyName` field to use some kind of reasonable default as a name in the function parameters, then. At the moment it looks like a bug, I feel.

mgermerie replied:
Okay, I'll add the property name as an argument then.

gkjohnson commented on src/utilities/BatchTable.js:
```diff
@@ -0,0 +1,68 @@
+		const extensions = this.header.extensions;
+		if ( extensions ) {
+
+			this.extensions = {};
```
We should always set `this.extensions = {}` even if there are no extensions. The class surface should be consistent, I think.

mgermerie replied:
You're right, I added it.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,122 @@
+
+		if ( extensionHeader.parentIds ) {
+
+			this.parentIds = this._parseProperty( extensionHeader.parentIds );
```
The `this._parseProperty` function takes a `propertyLength` but it's not passed in here. And if `parentIds` is always an array then there's no reason to pass it in to the function since the parseProperty doesn't do anything in that case.

It looks like [this isn't always an array, though](https://github.com/CesiumGS/3d-tiles/blob/87218a7c2030890b5200cf16cb466dbbd1480a20/extensions/3DTILES_batch_table_hierarchy/schema/3DTILES_batch_table_hierarchy.schema.json#L44-L46).

mgermerie replied:
Yes indeed, I added the length of `parentIds` to `_parseProperty` parameters.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,122 @@
+
+		if ( extensionHeader.parentIds ) {
+
+			this.parentIds = this._parseProperty( extensionHeader.parentIds );
```
Lets also make sure `parentIds` is set to a value regardless of whether it exists in the extension - even if that's an empty array or "null".

mgermerie replied:
I copied `classIds` as a default values for `parentIds`, since an instance whose parent is itself is considered as without parent (as mentioned [here](https://github.com/CesiumGS/3d-tiles/blob/87218a7c2030890b5200cf16cb466dbbd1480a20/extensions/3DTILES_batch_table_hierarchy/README.md?plain=1#L194)).

I get that we want it done to keep consistency in class properties, but be aware that this could cause issue in a specific case. Indeed, if user has data with a `parentCounts` different from an array of `1` and has no defined `parentIds`, this will create unwanted inheritance between instances that should not inherit any property. I get that this is a weird situation (having a `parentCounts` without `parentIds`) but that's not something prohibited by the specification.

gkjohnson replied:
I think it's okay to set `parentIds` to `null`, and that sounds like the better option given your explanation - I just wanted it to be set to _something_ in the general case so it's more clear what's going on when debugging.

mgermerie replied:
Good idea setting `parentIds` to `null`. I changed it.

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,122 @@
+
+		} else {
+
+			this.parentCounts = new Array( this.instancesLength ).fill( 1 );
```
Why default this to 1? Shouldn't it be defaulted to 0 if there are no parents specified?

mgermerie replied:
I agree it feels more logical but the specification defaults `parentCounts` to an array of ones ([here](https://github.com/CesiumGS/3d-tiles/blob/87218a7c2030890b5200cf16cb466dbbd1480a20/extensions/3DTILES_batch_table_hierarchy/README.md?plain=1#L190))

gkjohnson commented on src/utilities/BatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,122 @@
+
+			for ( let i = 0; i < parentCount; i ++ ) {
+
+				const parentId = this.parentIds[ parentIdsOffset + i ];
+				if ( parentId !== id ) {
+
+					Object.assign( target, this.getDataFromId( parentId ) );
+
+				}
```
Is there a reason to no just pass target into `this.getDataFromId` here, as well? We're assigning all the same values onto the object.

mgermerie replied:
That's an oversight from me. I added it.

gkjohnson reviewed (commented):
Just added a few more comments - thanks for bearing with me. I just want to make sure I understand the class and align the functions with the philosophies used in the rest of the project. I think it's just about there!

One last thing to do is update [this](https://github.com/NASA-AMMOS/3DTilesRendererJS?tab=readme-ov-file#read-batch-id-and-batch-table-data) section of the README to use `getDataFromId` instead of `getData`

mgermerie commented:
> Just added a few more comments - thanks for bearing with me. I just want to make sure I understand the class and align the functions with the philosophies used in the rest of the project. I think it's just about there!
> 
> One last thing to do is update [this](https://github.com/NASA-AMMOS/3DTilesRendererJS?tab=readme-ov-file#read-batch-id-and-batch-table-data) section of the README to use `getDataFromId` instead of `getData`

No problem, thanks for the reviews ! :wink: 
I added the modifications you suggested and updated the README.

gkjohnson commented on src/utilities/BatchTable.js:
```diff
@@ -0,0 +1,67 @@
+
+	getData( key, componentType = null, type = null ) {
+
+		console.warn( 'BatchTable: BatchTable.getData is deprecated. Use BatchTable.getDataFromId instead.' );
```
Forgot about the b3dmExample that tests batch tables - [these lines](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/a232cb76591f88241abc5e67ba4b91a95bc7d549/example/b3dmExample.js#L191-L194) should be changed now that we've changed the interface.

mgermerie replied:
Well spotted ! I updated the example.

gkjohnson reviewed (commented):
A couple final things but overall this looks good. Last thing - what's you're process for testing the changes so we can reference it in the future if changes need to be made?

mgermerie commented:
> A couple final things but overall this looks good. Last thing - what's you're process for testing the changes so we can reference it in the future if changes need to be made?

The testing process is not ideal : I tested the changes I made by plugging my local 3DTilesRendererJS branch to a local branch of [iTowns](https://github.com/iTowns/itowns) which I'll push soon (can't access my computer at the moment 😄). The local iTowns branch is basically [this one](https://github.com/iTowns/itowns/tree/3d-tiles-migration) where I added a call to `BatchTable.getDataFromId` in [this example](https://github.com/iTowns/itowns/blob/3d-tiles-migration/examples/3dtiles_batch_table.html) and logged its output.

I think I could as well add an example inspired by your [batch table example](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/a232cb76591f88241abc5e67ba4b91a95bc7d549/example/b3dmExample.js), replacing the dataset by the one used in iTowns example. It would be simpler and would have the same value in terms of testing quality.

That's only a functional test that is not testing all functionalities (as I mentioned I did not manage to find a dataset with binary batch table hierarchy extension), and I did not write proper unit tests.

gkjohnson commented:
> I think I could as well add an example inspired by your [batch table example](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/a232cb76591f88241abc5e67ba4b91a95bc7d549/example/b3dmExample.js), replacing the dataset by the one used in iTowns example. It would be simpler and would have the same value in terms of testing quality.

I don't think we need to add a new test page - just modifying that example to use a sample file with a batched hierarchy example is probably good enough as long as the basic "BatchedMesh" and highlighting functionality still works.

> That's only a functional test that is not testing all functionalities (as I mentioned I did not manage to find a dataset with binary batch table hierarchy extension), and I did not write proper unit tests.

That's okay. Writing unit tests for these kinds of things is difficult, I think. And it's even more difficult to test in general when we don't have any example data. It's okay to wait until we have user complaints and examples to address any issues.

mgermerie commented:
I updated the example to use a dataset using both batch table using hierarchy extension and a "classic" batch table. Would you rather have me squash all commits or keep the tree as it is now ?

gkjohnson commented:
Excellent! Thanks for all your work on this. We can squash commits when merging the PR.

mgermerie commented:
Thanks for the reviews :smiley:
